symengine/polys/uintpoly.cpp

```diff
@@ -1,7 +1,311 @@
+    for (auto &a : dict_) {
+        out += mp_abs(a.second);
+    }
+    return out;
+}
+
+UIntDict UIntDict::primitive() const
+{
+    integer_class gcd(0_z);
+    UIntDict out(dict_);
+    for (auto &a : dict_) {
+        mp_gcd(gcd, gcd, mp_abs(a.second));
+    }
+    for (auto &a : out.dict_) {
+        a.second /= gcd;
+    }
+    return out;
+}
+
+std::vector<UIntDict>
+UIntDict::zz_hensel_lift(const integer_class &p,
+                         const std::vector<UIntDict> &f_list,
+                         unsigned int l) const
+{
+    std::vector<UIntDict> res;
+    size_t r = f_list.size();
+    auto lc = get_lc();
+
+    if (r == 1) {
+        integer_class g, r, s, pl;
+        mp_pow_ui(pl, p, l);
+        mp_gcdext(g, r, s, lc, pl);
+        UIntDict F((*this) * r);
+        F.itrunc(pl);
+        res.push_back(F);
+        return res;
+    }
+
+    std::vector<UIntDict> sub1, sub2;
+    integer_class m(p);
+    size_t k(r / 2);
+    auto d = std::ceil(std::log2(l)); // TODO
+    auto g = GaloisFieldDict::from_vec({lc}, p);
+
+    for (unsigned i = 0; i < k; ++i) {
+        g *= GaloisFieldDict(f_list[i].dict_, p);
+        sub1.push_back(f_list[i]);
+    }
+
+    sub2.push_back(f_list[k]);
+    auto h = GaloisFieldDict(f_list[k].dict_, p);
+    for (unsigned i = k + 1; i < f_list.size(); ++i) {
+        h *= GaloisFieldDict(f_list[i].dict_, p);
+        sub2.push_back(f_list[i]);
+    }
+
+    GaloisFieldDict temp2, s, t;
+    GaloisFieldDict::gf_gcdex(g, h, outArg(s), outArg(t), outArg(temp2));
+
+    UIntDict dg = UIntDict::from_vec(g.dict_);
+    dg.itrunc(p);
+    UIntDict dh = UIntDict::from_vec(h.dict_);
+    dh.itrunc(p);
+    UIntDict ds = UIntDict::from_vec(s.dict_);
+    ds.itrunc(p);
+    UIntDict dt = UIntDict::from_vec(t.dict_);
+    dt.itrunc(p);
+
+    for (unsigned int i = 1; i <= d; ++i) {
+        UIntDict::zz_hensel_step(m, (*this), dg, dh, ds, dt, outArg(dg),
+                                 outArg(dh), outArg(ds), outArg(dt));
+        mp_pow_ui(m, m, 2);
+    }
+
+    res = dg.zz_hensel_lift(p, sub1, l);
+    auto temp = dh.zz_hensel_lift(p, sub2, l);
+    res.insert(res.end(), temp.begin(), temp.end());
+    return res;
+}
+
+void UIntDict::zz_divide(const UIntDict &a, const UIntDict &b,
+                         const Ptr<UIntDict> &quo, const Ptr<UIntDict> &rem)
+{
+    auto b_poly = b;
+    if (a.size() == 0)
+        throw std::runtime_error("Zero Division Error");
+    map_uint_mpz res;
+    UIntDict tmp;
+    integer_class q, r;
+    unsigned int a_deg, b_deg;
+    while (b_poly.degree() >= a.degree()) {
+        a_deg = a.degree();
+        b_deg = b_poly.degree();
+        q = b_poly.get_lc() / a.get_lc();
+        res[b_deg - a_deg] = q;
+        tmp = UIntDict(map_uint_mpz{{b_deg - a_deg, q}});
+        b_poly -= (a * tmp);
+    }
+    *quo = UIntDict(res);
+    *rem = b;
+    *rem -= (a * (*quo));
+}
+
+std::set<RCP<const UIntPoly>, RCPBasicKeyLess> UIntPoly::zz_zassenhaus() const
```

Comment:
This is a big method. Can you add a reference and also comment the code, so that I can understand what you are doing here?